File: UserClienJava/app/src/main/java/com/pacarius/userclienjava/MapTestPlot.java
package com.pacarius.userclienjava;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.Pair;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MapTestPlot extends View {
    private ImageView imageView;
    private int[] pngSize = new int[]{1670, 868};
    private int[] Vecs = new int[]{53, 28};
    private int[] imgPos;
    private CoordType[][] Grid = new CoordType[Vecs[0]][Vecs[1]];
    private List<Coordinates> vehicles = new ArrayList<>();
    private List<Coordinates> lampposts = new ArrayList<>();
    private ApiHandler handler = new ApiHandler(lampposts, vehicles);
    Pair<Float, Float> Size = new Pair<>(pngSize[0] / (float) Vecs[0], pngSize[1] / (float) Vecs[1]);
    @Override
    protected void onDraw(Canvas canvas){
        super.onDraw(canvas);
        getDimensions();
        drawAxes(canvas);
        drawSquares(canvas);
    }
    public MapTestPlot(Context context) {
        super(context);
        init();
    }
    public MapTestPlot(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }
    public MapTestPlot(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }
    private void init() {
        for (int i = 0; i < Grid.length; i++) {
            for (int j = 0; j < Grid[i].length; j++) {
                Grid[i][j] = CoordType.Empty;
            }
        }
        handler.setOnApiFire(() -> {
            for (int i = 0; i < Grid.length; i++) {
                for (int j = 0; j < Grid[i].length; j++) {
                    Grid[i][j] = CoordType.Empty;
                }
            }
            for (Coordinates coord : lampposts) {
                Grid[coord.getX()][coord.getY()] = CoordType.Lamppost;
            }
            for (Coordinates coord : vehicles) {
                Grid[coord.getX()][coord.getY()] = CoordType.Vehicle;
            }
            redraw();
        });
    }

    public void redraw() {
        invalidate(); // Redraw the view
    }
    public ApiHandler handler() {
        return handler;
    }
    private void getDimensions() {
        Drawable drawable = imageView.getDrawable();
        if (drawable != null) {
            int intrinsicWidth = drawable.getIntrinsicWidth();
            int intrinsicHeight = drawable.getIntrinsicHeight();
            int imageViewWidth = imageView.getWidth() - imageView.getPaddingLeft() - imageView.getPaddingRight();
            int imageViewHeight = imageView.getHeight() - imageView.getPaddingTop() - imageView.getPaddingBottom();
            float scaleFactor = Math.min(
                    imageViewWidth / (float) intrinsicWidth,
                    imageViewHeight / (float) intrinsicHeight
            );
            int imageWidth = (int) (intrinsicWidth * scaleFactor);
            int imageHeight = (int) (intrinsicHeight * scaleFactor);
            int imageLeft = (imageViewWidth - imageWidth) / 2 + imageView.getPaddingLeft();
            int imageTop = (imageViewHeight - imageHeight) / 2 + imageView.getPaddingTop();
            setDimensions(new int[]{imageWidth, imageHeight});
            setPosition(new int[]{imageLeft, imageTop});
        }
    }

    private void setDimensions(int[] a) {
        pngSize[0] = Math.max(a[0], a[1]);
        pngSize[1] = Math.min(a[0], a[1]);
    }

    private void setPosition(int[] a) {
        imgPos = a;
    }

    private void drawAxes(Canvas canvas) {
        Paint paint = new Paint();
        paint.setColor(Color.BLACK);
        paint.setStrokeWidth(2f);
        float padding = paint.getStrokeWidth() / 2;

        for (int x = 0; x < Vecs[0] + 1; x++) {
            for (int y = 0; y < Vecs[1] + 1; y++) {
                int posX = imgPos[0] + x * pngSize[0] / Vecs[0];
                int posY = imgPos[1] + y * pngSize[1] / Vecs[1];
                if (x < Vecs[0]) {
                    canvas.drawLine(posX, posY, posX + pngSize[0] / Vecs[0], posY, paint);
                }
                if (y < Vecs[1]) {
                    canvas.drawLine(posX, posY, posX, posY + pngSize[1] / Vecs[1], paint);
                }
            }
        }
    }

    private void drawSquares(Canvas canvas) {
        for (int x = 0; x < Grid.length; x++) {
            for (int y = 0; y < Grid[x].length; y++) {
                if (Grid[x][y] != CoordType.Empty) {
                    int color;
                    switch (Grid[x][y]) {
                        case Lamppost:
                            color = Color.RED;
                            break;
                        case Vehicle:
                            color = Color.CYAN;
                            break;
                        default:
                            color = Color.WHITE;
                            break;
                    }
                    float left = imgPos[0] + x * pngSize[0] / (float) Vecs[0];
                    float top = imgPos[1] + (Vecs[1] - y + 1) * pngSize[1] / (float) Vecs[1];
                    float right = imgPos[0] + (x + 1) * pngSize[0] / (float) Vecs[0];
                    float bottom = imgPos[1] + (Vecs[1] - y) * pngSize[1] / (float) Vecs[1];
                    Paint paint = new Paint();
                    paint.setColor(color);
                    canvas.drawRect(left, top, right, bottom, paint);
                }
            }
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            float x = event.getX();
            float y = getHeight() - event.getY();

            int gridX = (int) ((x - imgPos[0]) / (pngSize[0] / (float) Vecs[0]));
            int gridY = (int) ((y - imgPos[1]) / (pngSize[1] / (float) Vecs[1]));

            if (gridX >= 0 && gridX < Vecs[0] && gridY >= 0 && gridY < Vecs[1]) {
                CoordType cellContent = Grid[gridX][gridY];
                String tmp = "Square at: (" + gridX + ", " + gridY + ") contains: ";
                if (cellContent != CoordType.Empty) {
                    Toast.makeText(getContext(), tmp + cellContent, Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getContext(), tmp + "Nothing", Toast.LENGTH_SHORT).show();
                }
                return true;
            }
        }
        return super.onTouchEvent(event);
    }

    public void setImageView(ImageView iv) {
        imageView = iv;
    }
}

File: UserClient/app/src/main/java/com/pacarius/client/MapTestPlot.kt
package com.pacarius.client

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.media.Image
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.Toast
import kotlin.math.min

class MapTestPlot : View {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )
    public var imageView: ImageView? = null
    private val pngSize = intArrayOf(1670, 868)
    private val Vecs = intArrayOf(53, 28)
    private var imgPos: IntArray? = null
    public val Grid = Array(Vecs[0]) { Array(Vecs[1]) { CoordType.Empty } }
    val vehicles = mutableListOf<Coordinates>()
    val lampposts = mutableListOf<Coordinates>()
    public val handler = ApiHandler(lampposts, vehicles)
    val Size
        get() = floatArrayOf(pngSize[0] / Vecs[0].toFloat(), pngSize[1] / Vecs[1].toFloat())
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        getDimensions()
        drawAxes(canvas)
        drawSquares(canvas)
    }
    init{
        Grid[10][17] = CoordType.Lamppost
        Grid[46][17] = CoordType.Lamppost
        handler.OnApiFire = {
            for (i in Grid) {
                for (index in i.indices) {
                    i[index] = CoordType.Empty
                }
            }
            for (i in lampposts) {
                Grid[i.x][i.y] = CoordType.Lamppost
            }
            for (i in vehicles) {
                Grid[i.x][i.y] = CoordType.Vehicle
            }
            redraw()
        }
    }
    fun redraw() {
        invalidate() // Redraw the view
    }
    private fun getDimensions() {
        val drawable = imageView!!.drawable
        drawable.let {
            val intrinsicWidth = it.intrinsicWidth
            val intrinsicHeight = it.intrinsicHeight
            val imageViewWidth =
                imageView!!.width - imageView!!.paddingLeft - imageView!!.paddingRight
            val imageViewHeight =
                imageView!!.height - imageView!!.paddingTop - imageView!!.paddingBottom
            val scaleFactor = min(
                imageViewWidth / intrinsicWidth.toFloat(),
                imageViewHeight / intrinsicHeight.toFloat()
            )
            val imageWidth = intrinsicWidth * scaleFactor
            val imageHeight = intrinsicHeight * scaleFactor
            val imageLeft = (imageViewWidth - imageWidth) / 2 + imageView!!.paddingLeft
            val imageTop = (imageViewHeight - imageHeight) / 2 + imageView!!.paddingTop
            setDimensions(Pair(imageWidth.toInt(), imageHeight.toInt()))
            setPosition(Pair(imageLeft.toInt(), imageTop.toInt()))
        }
    }
    private fun setDimensions(a: Pair<Int, Int>) {
        val tmp = intArrayOf(a.first, a.second)
        pngSize[0] = tmp.max()
        pngSize[1] = tmp.min()
    }
    private fun setPosition(a: Pair<Int, Int>) {
        val tmp = intArrayOf(a.first, a.second)
        imgPos = tmp
    }
    private fun drawAxes(canvas: Canvas) {
    val paint = Paint()
    paint.color = Color.BLACK
    paint.strokeWidth = 2f
    // Padding to ensure lines are fully visible
    val padding = paint.strokeWidth / 2

    // Draw axes at the edges of each grid cell
    for (x in 0 until Vecs[0] + 1) {
        for (y in 0 until Vecs[1] + 1) {
            val posX = imgPos!![0] + x * pngSize[0] / Vecs[0]
            val posY = imgPos!![1] + y * pngSize[1] / Vecs[1]
            if (x < Vecs[0]) {
                // Draw horizontal grid lines
                canvas.drawLine(posX.toFloat(), posY.toFloat(),
                    (posX + pngSize[0] / Vecs[0]).toFloat(), posY.toFloat(), paint)
            }
            if (y < Vecs[1]) {
                // Draw vertical grid lines
                canvas.drawLine(posX.toFloat(), posY.toFloat(),
                    posX.toFloat(), (posY + pngSize[1] / Vecs[1]).toFloat(), paint)
            }
        }
    }
}
    private fun drawSquares(canvas: Canvas){
    for((x, i)in Grid.withIndex()){
        for((y, j)in i.withIndex()){
            if(j != CoordType.Empty){
                val color = when(j){
                    CoordType.Lamppost -> Color.RED
                    CoordType.Vehicle -> Color.CYAN
                    else -> Color.WHITE
                }
                val left = imgPos!![0] + x * pngSize[0] / Vecs[0].toFloat()
                val top = imgPos!![1] + (Vecs[1] - y + 1) * pngSize[1] / Vecs[1].toFloat() // Invert the y-coordinate
                val right = imgPos!![0] + (x + 1) * pngSize[0] / Vecs[0].toFloat()
                val bottom = imgPos!![1] + (Vecs[1] - y) * pngSize[1] / Vecs[1].toFloat() // Invert the y-coordinate
                val paint = Paint()
                paint.color = color
                canvas.drawRect(
                    left,
                    top,
                    right,
                    bottom,
                    paint
                )
            }
        }
    }
}
    override fun onTouchEvent(event: MotionEvent): Boolean {
    if (event.action == MotionEvent.ACTION_DOWN) {
        val x = event.x
        val y = height - event.y // Flip the y-coordinate

        // Calculate the grid cell that the user has tapped on
        val gridX = ((x - imgPos!![0]) / (pngSize[0] / Vecs[0])).toInt()
        val gridY = ((y - imgPos!![1]) / (pngSize[1] / Vecs[1])).toInt()

        // Check if the tap event is within the grid
        if (gridX in 0 until Vecs[0] && gridY in 0 until Vecs[1]) {
            val cellContent = Grid[gridX][gridY]
            val tmp = "Square at: ($gridX, $gridY) contains: "
            if(cellContent != CoordType.Empty){
                Toast.makeText(context, tmp + cellContent, Toast.LENGTH_SHORT).show()
            }
            else{
                Toast.makeText(context, tmp + "Nothing", Toast.LENGTH_SHORT).show()
            }
            return true
        }
    }
    return super.onTouchEvent(event)
}
}